### PR TITLE
FINERACT-2081: Fix reversal of loan accruals when loan repayment got reversed

### PR DESCRIPTION
**Pull request by alberto-art3ch**
## Description

Due the Loan accrual transaction currently is not added when the charge is added, there is an issue when It is added after some time and a loan repayment transaction is reversed 

[FINERACT-2081](https://issues.apache.org/jira/browse/FINERACT-2081)

Screenshot 2024-10-15 at 4 09 30 p m

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanChargeWritePlatformServiceImpl.java [hunk 1]
```diff
@@ -1041,7 +1041,8 @@ private boolean addCharge(final Loan loan, final Charge chargeDefinition, LoanCh
 
         // we want to apply charge transactions only for those loans charges that are applied when a loan is active and
         // the loan product uses Upfront Accruals
-        if (loan.getStatus().isActive() && loan.isNoneOrCashOrUpfrontAccrualAccountingEnabledOnLoanProduct()) {
+        if ((loan.getStatus().isActive() && loan.isNoneOrCashOrUpfrontAccrualAccountingEnabledOnLoanProduct())
+                || (loan.getStatus().isOverpaid() && loan.isPeriodicAccrualAccountingEnabledOnLoanProduct())) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanChargeWritePlatformServiceImpl.java]: Would you mind to change the condition to:
- Allow to create the accrual transaction if the loan is closed? I dont think we need the "is periodic accrual enabled" and also CLOSED_OBLIGATIONS_MET is a valid outcome after the charge was added:
- It was overpaid but charge was added and now it is closed (edge case, but why not...)

Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanChargeWritePlatformServiceImpl.java]: Done! conditions added!

**Review by alberto-art3ch — COMMENTED** [review]
Done!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 2]
```diff
@@ -5275,6 +5277,71 @@ public void uc148c() {
...
+
+            addRepaymentForLoan(createdLoanId.get(), 600.00, operationDate);
+
+            createdLoanChargeId.set(addCharge(createdLoanId.get(), true, 30, operationDate));
+
+            executeInlineCOB(createdLoanId.get());
+        });
+
+        runAt("10 October 2024", () -> {
+            PostLoansLoanIdChargesChargeIdRequest request = new PostLoansLoanIdChargesChargeIdRequest().amount(15.0).locale("en");
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java:5336]: Would you mind checking whether the accrual was not reverted?

Comment (alberto-art3ch) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java:5336]: Done! Test case updated

**Review by alberto-art3ch — COMMENTED** [review]
Completed

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**